python/python/lance/arrow.py

```diff
@@ -11,14 +11,347 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-from typing import Optional, Union
+from pathlib import Path
+from typing import Iterable, Optional, Union
 
 import pyarrow as pa
 
 from .lance import BFloat16
 from .lance import bfloat16_array as bfloat16_array
 
 
+# TODO: add VariableShapeImageTensorType once pa.VariableShapeTensorArray is available
+class ImageURIType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.string(), "lance.arrow.image_uri")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageURIType()
+
+    def __arrow_ext_class__(self):
+        return ImageURIArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageURIScalar
+
+
+class EncodedImageType(pa.ExtensionType):
+    def __init__(self):
+        # TODO: use pa.BinaryView once available?
+        pa.ExtensionType.__init__(self, pa.binary(), "lance.arrow.encoded_image")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return EncodedImageType()
+
+    def __arrow_ext_class__(self):
+        return EncodedImageArray
+
+    def __arrow_ext_scalar_class__(self):
+        return EncodedImageScalar
+
+
+class FixedShapeImageTensorType(pa.ExtensionType):
+    def __init__(self, arrow_type, shape):
+        pa.ExtensionType.__init__(
+            self,
+            pa.fixed_shape_tensor(arrow_type, shape),
+            "lance.arrow.fixed_shape_image_tensor",
+        )
+
+    def __arrow_ext_serialize__(self):
+        return b""
```

Comment:
Related to the question of subclassing FixedShapeTensor: shouldn't we be encoding the shape and such in here?

Could we add a test where we write each of these array types to Lance and verify we get the same array back?

Reply:
I switched `FixedShapeImageTensorArray` storage to `pa.FixedSizeListArray` and added relevant metadata serialization/deserialization. However it seems that in roundtrip deserialization step `pa.FixedSizeListArray.__deserialize__` (on storage type) is called instead of `pa.FixedShapeImageTensorType.__deserialize__` (on extension type).